src/pybind/mgr/cephadm/services/cephadmservice.py

```diff
@@ -989,3 +990,38 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
         daemon_spec.keyring = keyring
         daemon_spec.final_config, daemon_spec.deps = self.generate_config(daemon_spec)
         return daemon_spec
+
+
+class CephadmAgent(CephService):
+    TYPE = 'agent'
+
+    def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonDeploySpec:
+        assert self.TYPE == daemon_spec.daemon_type
+        daemon_id, host = daemon_spec.daemon_id, daemon_spec.host
+
+        daemon_spec.ports = [self.mgr.agent_port]
+
+        keyring = self.get_keyring_with_caps(self.get_auth_entity(daemon_id, host=host), [])
+        daemon_spec.keyring = keyring
+        self.mgr.cache.agent_keys[host] = keyring
+
+        daemon_spec.final_config, daemon_spec.deps = self.generate_config(daemon_spec)
+
+        return daemon_spec
+
+    def generate_config(self, daemon_spec: CephadmDaemonDeploySpec) -> Tuple[Dict[str, Any], List[str]]:
+        cfg = {'target_ip': self.mgr.get_mgr_ip(),
+               'target_port': self.mgr.endpoint_port,
+               'refresh_period': self.mgr.agent_refresh_rate,
+               'listener_port': self.mgr.agent_port}
+
+        config = {
+            'agent.json': json.dumps(cfg),
+            'cephadm': self.mgr._cephadm,
+            'keyring': daemon_spec.keyring
+        }
+
+        self.mgr.cache.agent_given_ip_port[daemon_spec.host] = (
+            self.mgr.get_mgr_ip(), self.mgr.endpoint_port)
+
+        return config, []
```

Comment:
That second return list can describe what teh config dependency was.  in our case, I would make it the current mgr ip + port.  Then the existing mgr reconfig machinery will handle redeploying agents after a mgr failover...